controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java

```diff
@@ -0,0 +1,417 @@
+         * replaces older one. <em>Note:</em> {@link CredentialStoreSpi} supports storing of multiple entries (credential types) per alias.
+         * Each must be of different credential type.
+         *
+         * @param credentialAlias to store the credential to the store
+         * @param credential      instance of {@link Credential} to store
+         * @throws CredentialStoreException           when the credential cannot be stored
+         * @throws UnsupportedCredentialTypeException when the credentialType is not supported
+         */
+        @Override
+        public <C extends Credential> void store(String credentialAlias, C credential) throws CredentialStoreException, UnsupportedCredentialTypeException {
+            throw new RuntimeException("method not implemented");
+        }
+
+        /**
+         * Retrieve credential stored in the store under the key and of the credential type
+         *
+         * @param credentialAlias to find the credential in the store
+         * @param credentialType  - credential type to retrieve from under the credentialAlias from the store
+         * @return instance of {@link Credential} stored in the store
+         * @throws CredentialStoreException           - if credentialAlias credentialType combination doesn't exist or credentialAlias cannot be retrieved
+         * @throws UnsupportedCredentialTypeException when the credentialType is not supported
+         */
+        @Override
+        public <C extends Credential> C retrieve(String credentialAlias, Class<C> credentialType) throws CredentialStoreException, UnsupportedCredentialTypeException {
+            try {
+                PasswordFactory passwordFactory = PasswordFactory.getInstance(ClearPassword.ALGORITHM_CLEAR);
+                return credentialType.cast(new PasswordCredential(passwordFactory.generatePassword(new ClearPasswordSpec(credentialReference.getSecret()))));
+            } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+                throw new CredentialStoreException(e);
+            }
+        }
+
+        /**
+         * Remove the credentialType with from given alias from the store.
+         *
+         * @param credentialAlias alias to remove
+         * @param credentialType  - credential type to be removed from under the credentialAlias from the store
+         * @throws CredentialStoreException           - if credentialAlias credentialType combination doesn't exist or credentialAlias cannot be removed
+         * @throws UnsupportedCredentialTypeException when the credentialType is not supported
+         */
+        @Override
+        public <C extends Credential> void remove(String credentialAlias, Class<C> credentialType) throws CredentialStoreException, UnsupportedCredentialTypeException {
+            throw new RuntimeException("method not implemented");
```

Comment:
Should this be one of the 2 documented exception types?

I only care if this class isn't private.